### PR TITLE
Refactor registration activity to UI/ViewModel/Model layers

### PR DESCRIPTION
**Pull request by pmcintyr**
refactor the registerUser class into UI/ViewModel/Model as seen in the App's architecture in issue https://github.com/BrugApp/brug/issues/98.

### CODE REVIEW

File: app/src/main/java/com/github/brugapp/brug/ui/RegisterUserActivity.kt [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_register_user)
+        val regUser = findViewById<Button>(R.id.registerbutton)
+        regUser.setOnClickListener(this)
+        progressBar = findViewById(R.id.progressBar)
+    }
+
+    override fun onClick(v: View?) { //we clicked the register button
+        viewModel.storeUserInput(
+            FirebaseAuth.getInstance(),
```
Comment (DeepGreen1) [inline app/src/main/java/com/github/brugapp/brug/ui/RegisterUserActivity.kt:34]: FirebaseAuth.getInstance() is java code, you could update it for kotlin code: Firebase.auth (see here: https://firebase.google.com/docs/auth/android/start#kotlin+ktx_2 )

Also, be careful with the next line (35) that I think, would return the same database (kotlin code this time)

File: app/src/main/java/com/github/brugapp/brug/ui/RegisterUserActivity.kt [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        if (!viewModel.anyEmpty()) {
+            onClickHelper()
+        }
+    }
+
+    private fun onClickHelper() {
+        progressBar.visibility = View.VISIBLE
+        //make an authentication account with email password tuple
+        viewModel.createAuthAccount()
+            .addOnCompleteListener(this) { task ->
```
Comment (DeepGreen1) [inline app/src/main/java/com/github/brugapp/brug/ui/RegisterUserActivity.kt:50]: To have even better separation of UI/Logic, you could try to split this method. If it's possible, the createAuthAccount() should handle all the logic that is made here: call addRegisterUserTask and print log text where the onClickHelper should only update the progressBar, show the toast and call the method inside the viewmodel.

File: app/src/main/java/com/github/brugapp/brug/view_model/RegisterUserViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+import android.widget.ProgressBar
+import androidx.lifecycle.ViewModel
+import com.google.android.gms.tasks.Task
+import com.google.firebase.auth.FirebaseAuth
+import com.google.firebase.firestore.DocumentReference
+import com.google.firebase.firestore.FirebaseFirestore
+
+class RegisterUserViewModel : ViewModel() {
+    private lateinit var db: FirebaseFirestore
+    private lateinit var progressBar: ProgressBar
```
Comment (DeepGreen1) [inline app/src/main/java/com/github/brugapp/brug/view_model/RegisterUserViewModel.kt:13]: To separate view/Logic, the viewmodel should not contain any UI element (progressBar / EditText fields). You can send these values as parameter when you call the method from the activity.

File: app/src/main/java/com/github/brugapp/brug/view_model/RegisterUserViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+                email.requestFocus()
+            }
+            else -> {
+                return false
+            }
+        }
+        return true
+    }
+
+    fun addRegisterUserTask(userToAdd: HashMap<String, Any>): Task<DocumentReference> {
```
Comment (DeepGreen1) [inline app/src/main/java/com/github/brugapp/brug/view_model/RegisterUserViewModel.kt:93]: Maybe add a TODO and/or a task in the scrum board to think to update this part of the code when the database class is created. (db will be changed from FirebaseFirestore to our new fancy class :) )

**Review by DeepGreen1 — DISMISSED** [review]
Good work for the refactoring!
I let you some comments to read; l can approve the request if we agree to do them with the next refactor (the one due to the new "database" class), as we are all depending on the architecture changes before implementing the actual weekly work.

**Review by ha2san — APPROVED** [review]
Really nice work, I have nothing to add to what @DeepGreen1 said. I'm not an expert in app architecture but it seems adequate to me.

File: app/src/main/java/com/github/brugapp/brug/view_model/RegisterUserViewModel.kt (hunk 3 above)
Comment (pmcintyr) [reply app/src/main/java/com/github/brugapp/brug/view_model/RegisterUserViewModel.kt]: This causes lots of **code duplication** when passing the same parameters to multiple functions but i did it anyway. If this issue comes up, I will redirect them here.

File: app/src/main/java/com/github/brugapp/brug/ui/RegisterUserActivity.kt (hunk 2 above)
Comment (pmcintyr) [reply app/src/main/java/com/github/brugapp/brug/ui/RegisterUserActivity.kt]: done

File: app/src/main/java/com/github/brugapp/brug/ui/RegisterUserActivity.kt (hunk 1 above)
Comment (pmcintyr) [reply app/src/main/java/com/github/brugapp/brug/ui/RegisterUserActivity.kt]: done in https://github.com/BrugApp/brug/pull/110/commits/d326ba81c162987a9633c1da83820ccc7995397b

File: app/src/main/java/com/github/brugapp/brug/ui/RegisterUserActivity.kt [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        setContentView(R.layout.activity_register_user)
+        val regUser = findViewById<Button>(R.id.registerbutton)
+        regUser.setOnClickListener(this)
+        progressBar = findViewById(R.id.progressBar)
+    }
+
+    override fun onClick(v: View?) { //we clicked the register button
+        viewModel.storeUserInput(
+            Firebase.auth,
+            Firebase.firestore,
```
Comment (omarelmalki) [inline app/src/main/java/com/github/brugapp/brug/ui/RegisterUserActivity.kt:36]: Here, the Activity layer (Activity class + xml file) should not be aware that the viewModel uses Firebase. I suggest keeping Firebase.auth and Firebase.firestore as attributes in the RegisterUserViewModel class so you do not have to put them as argument.
```suggestion
```

File: app/src/main/java/com/github/brugapp/brug/view_model/RegisterUserViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+import android.widget.Toast
+import androidx.lifecycle.ViewModel
+import com.google.android.gms.tasks.Task
+import com.google.firebase.auth.FirebaseAuth
+import com.google.firebase.firestore.DocumentReference
+import com.google.firebase.firestore.FirebaseFirestore
+
+class RegisterUserViewModel : ViewModel() {
+    private lateinit var db: FirebaseFirestore
+    private lateinit var mAuth: FirebaseAuth
```
Comment (omarelmalki) [inline app/src/main/java/com/github/brugapp/brug/view_model/RegisterUserViewModel.kt:17]:
```suggestion
    private val db: FirebaseFirestore = Firebase.firestore
    private val mAuth: FirebaseAuth = Firebase.auth
```

File: app/src/main/java/com/github/brugapp/brug/view_model/RegisterUserViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+            "firstName" to firstnametxt,
+            "lastName" to lastnametxt
+        )
+        return userToAdd
+    }
+
+    //stores user input data in the view model
+    fun storeUserInput(
+        mAuth: FirebaseAuth,
+        db: FirebaseFirestore,
```
Comment (omarelmalki) [inline app/src/main/java/com/github/brugapp/brug/view_model/RegisterUserViewModel.kt:40]:
```suggestion
```

File: app/src/main/java/com/github/brugapp/brug/view_model/RegisterUserViewModel.kt [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+        firstName: EditText,
+        lastName: EditText,
+        password: EditText,
+        email: EditText
+    ) {
+        this.db = db
+        this.mAuth = mAuth
```
Comment (omarelmalki) [inline app/src/main/java/com/github/brugapp/brug/view_model/RegisterUserViewModel.kt:47]:
```suggestion
```

**Review by omarelmalki — CHANGES_REQUESTED** [review]
Great job on the refactoring !
Make sure to preserve the abstraction between each layer. Checkout my suggestions to make sure the activity does not know the rest of the app is using Firebase.

File: app/src/main/java/com/github/brugapp/brug/ui/RegisterUserActivity.kt (hunk 5 above)
Comment (pmcintyr) [reply app/src/main/java/com/github/brugapp/brug/ui/RegisterUserActivity.kt]: good idea! thank you for this feedback, i have now implemented it in this commit: https://github.com/BrugApp/brug/pull/110/commits/4188a218e6f70b8aae558fc29f648dce21812b74

**Pull request merged**